src/main/java/net/snowflake/ingest/connection/OAuthManager.java

```diff
@@ -0,0 +1,177 @@
+      String accountName,
+      String username,
+      OAuthCredential oAuthCredential,
+      URIBuilder baseURIBuilder,
+      double updateThresholdRatio,
+      TelemetryService telemetryService) {
+    // disable telemetry service until jdbc v3.13.34 is released
+    super(accountName, username, null);
+
+    // if any of our arguments are null, throw an exception
+    if (oAuthCredential == null || baseURIBuilder == null) {
+      throw new IllegalArgumentException();
+    }
+
+    // check if update threshold is within (0, 1)
+    if (updateThresholdRatio <= 0 || updateThresholdRatio >= 1) {
+      throw new IllegalArgumentException("updateThresholdRatio should fall in (0, 1)");
+    }
+    this.updateThresholdRatio = updateThresholdRatio;
+    this.oAuthClient = new SnowflakeOAuthClient(accountName, oAuthCredential, baseURIBuilder);
+
+    // generate our first token
+    refreshToken();
+  }
+
+  /**
+   * Creates a OAuthManager entity for a given account, user and OAuthClient with a specified time
+   * to refresh the token. Use for testing only.
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param oAuthClient - the OAuth client to perform token refresh
+   * @param updateThresholdRatio - the ratio between the expiration time of a token and the time
+   *     needed to refresh it.
+   */
+  public OAuthManager(
+      String accountName, String username, OAuthClient oAuthClient, double updateThresholdRatio) {
+    super(accountName, username, null);
+
+    this.updateThresholdRatio = updateThresholdRatio;
+    this.oAuthClient = oAuthClient;
+
+    refreshToken();
+  }
+
+  @Override
+  String getToken() {
+    if (refreshFailed.get()) {
+      LOGGER.error("getToken request failed due to token refresh failure");
+      throw new SecurityException();
+    }
+    return oAuthClient.getoAuthCredentialRef().get().getAccessToken();
+  }
+
+  @Override
+  String getTokenType() {
+    return TOKEN_TYPE;
+  }
+
+  /**
+   * Set refresh token, this method is for refresh token renewal without requiring to restart
+   * client. This method only works when the authorization type is OAuth.
+   *
+   * @param refreshToken the new refresh token
+   */
+  void setRefreshToken(String refreshToken) {
+    oAuthClient.getoAuthCredentialRef().get().setRefreshToken(refreshToken);
+  }
+
+  /** refreshToken - Get new access token using refresh_token, client_id, client_secret */
+  private void refreshToken() {
+    for (int retries = 0; retries < Constants.MAX_REFRESH_TOKEN_RETRY; retries++) {
+      try {
+        oAuthClient.refreshToken();
+
+        // Schedule next refresh
+        long nextRefreshDelay =
+            (long)
+                (oAuthClient.getoAuthCredentialRef().get().getExpires_in()
+                    * this.updateThresholdRatio);
+        tokenRefresher.schedule(this::refreshToken, nextRefreshDelay, TimeUnit.SECONDS);
```

Comment:
Where is the tokenRefresher gets initialized?

Reply:
In the [super class](https://github.com/snowflakedb/snowflake-ingest-java/blob/690c3b95fc0cc09a857cb244ce05b46b91e90dc2/src/main/java/net/snowflake/ingest/connection/SecurityManager.java#L39C1-L39C1)